Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -484,6 +517,39 @@ public CsharpCompiler(ErrorReporter reporter)
       }
     }
 
+    public override bool NeedsCustomReceiver(MemberDecl member) {
+      Contract.Requires(member != null);
+      if (!member.IsStatic && member.EnclosingClass is NewtypeDecl) {
+        return true;
+      }
+      if (!member.IsStatic && member.EnclosingClass is IndDatatypeDecl d) {
+        if (member is Function f) {
```

Comment:
Probably needs the same for Methods. Will add a test.

Reply:
That function generally needs a second thought, concerning subtypes and ghostness.